File: cpp_solution/TripService.cpp
#include "pch.h"
#include "TripService.h"

#include <list>


User* TripService::getLoggedInUser()
{
	return UserSession::GetInstance()->GetLoggedUser();
}

std::list<Trip> TripService::findTripsByUser(User& user)
{
	return TripDAO::FindTripsByUser(user);
}


std::list<Trip> TripService::GetTripsByUser(User& user, User* loggedUser)
{
	User* loggedUserLocal = getLoggedInUser();

	if (!loggedUserLocal)
	{
		throw UserNotLoggedInException("UserNotLoggedInException");
	}

	if (user.isFriendWith(*loggedUserLocal))
	{
		return findTripsByUser(user);
	}

	return NO_TRIPS;
}

File: c#/TripServiceKata/TripServiceKata/Trip/TripService.cs
using System.Collections.Generic;
using TripServiceKata.Exception;
using TripServiceKata.User;

namespace TripServiceKata.Trip
{
    public class TripService
    {
        private readonly IUserSession _userSession;
        private readonly ITripRepository _tripRepository;

        public TripService(IUserSession userSession, ITripRepository tripRepository)
        {
            _userSession = userSession;
            _tripRepository = tripRepository;
        }

        public List<Trip> GetTripsByUser(User.User user)
        {
            User.User loggedUser = _userSession.GetLoggedUser();
            if (loggedUser == null)
            {
                throw new UserNotLoggedInException();
            }
            return user.IsFriend(loggedUser)
                ? _tripRepository.FindTripsByUser(user)
                : new List<Trip>();
        }
    }
}
